### PR TITLE
Add documentation for `FileService` & `FileSystemProvider`

### LINKED ISSUE #8599: Add Documentation for FileService API
The`FileService` and related concepts (like. `FileSystemProvider`) are not well documented. Augmenting the existing documentation would help users to better understand how to interact with and contribute to the `FileService`.
I will create a PR.

### PR DESCRIPTION
**Pull request by tortmayr**
Fixes #8599 
Contributed on behalf of STMicroelectronics
Signed-off-by: Tobias Ortmayr <tortmayr@eclipsesource.com>

#### What it does
Add documentation for core components of the  `FileService` & `FileSystemProvider` API
#### How to test
Documentation changes only
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/filesystem/src/common/files.ts [hunk 1]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+     * Create a new directory using the given resource uri.
+     * @param resource The `URI` of the new folder.
+     */
     mkdir(resource: URI): Promise<void>;
+
+    /**
+     * Retrieve the content of a given directory.
+     * @param resource The `URI` of the directory.
+     *
+     * @returns A map containg the {@link FileType} for each child resource (uri).
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:609]:
```suggestion
     * @returns A map containing the {@link FileType} for each child resource (uri).
```

File: packages/filesystem/src/common/files.ts [hunk 2]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
 }
 
 export const FileSystemProvider = Symbol('FileSystemProvider');
+/**
+ * A {@link FileSystemProvider} provides the capabilities to read, write, discover, and to mangage files and folders
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:562]:
```suggestion
 * A {@link FileSystemProvider} provides the capabilities to read, write, discover, and to manage files and folders
```

File: packages/filesystem/src/common/files.ts [hunk 3]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+    /** Event that is fired if a (watched) file in the filesystem of this provider has changed. */
     readonly onDidChangeFile: Event<readonly FileChange[]>;
+
+    /** Event that is fired if an error occurred when watching files in the filesystem of this provider. */
     readonly onFileWatchError: Event<void>;
+
+    /**
+     * Watch the given resource and react to changes by firing the {@link FileSystemProvider#onDidChangeFile} event.
+     * @param resource `URI` of the resource to be watched.
+     * @param opts Options to define if the resource should be watched recursivly and to
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:584]:
```suggestion
     * @param opts Options to define if the resource should be watched recursively and to
```

File: packages/filesystem/src/common/files.ts [hunk 4]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+     * @param resource The `URI` of the directory.
+     *
+     * @returns A map containg the {@link FileType} for each child resource (uri).
+     */
     readdir(resource: URI): Promise<[string, FileType][]>;
+
+    /**
+     * Delete the given resource.
+     * @param resource The `URI` of the resource to delete.
+     * @param opts Options to define if files should be deleted recursivly and if the trash should be used.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:616]:
```suggestion
     * @param opts Options to define if files should be deleted recursively and if the trash should be used.
```

File: packages/filesystem/src/common/files.ts [hunk 5]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
 }
 
+/**
+ * Subtype of {@link FileSystemProvider} that ensures that the optional functions needed for providers, that should be
+ * able access files, are implemented.
+ */
 export interface FileSystemProviderWithAccessCapability extends FileSystemProvider {
+    /**
+     * Test if the user has the permission to access the given file in the specified mode.
+     * @param resource The `URI` of the file that should be testet.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:703]:
```suggestion
     * @param resource The `URI` of the file that should be tested.
```

File: packages/filesystem/src/browser/file-service.ts [hunk 6]
```diff
@@ -174,37 +174,72 @@ export interface UserFileOperationEvent extends WaitUntilEvent {
...
     added: boolean;
+    /** The (uri) scheme for which the provider was (previously) registered */
     scheme: string;
+    /** The affected file system provider for which this event was fired. */
     provider?: FileSystemProvider;
 }
 
+/**
+ * Represents the `FileSystemProviderCapabilitiesChange` event.
+ * This event is fired by the {@link FileService} if the capabilites of one of its managed
```
Comment (vince-fugnitto) [inline packages/filesystem/src/browser/file-service.ts:226]:
```suggestion
 * This event is fired by the {@link FileService} if the capabilities of one of its managed
```

File: packages/filesystem/src/browser/file-service.ts [hunk 7]
```diff
@@ -394,6 +461,13 @@ export class FileService {
      */
     readonly onDidRunOperation = this.onDidRunOperationEmitter.event;
 
+    /**
+     * Try to resolve file information and metadata for the given resource.
+     * @param resource `URI` of the resource that should be resolved.
+     * @param options  Options to customize the resolvment process.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/browser/file-service.ts:467]:
```suggestion
     * @param options  Options to customize the resolved process.
```

If `resolvement` is the word we want here then I believe this should be spelling.

File: packages/filesystem/src/browser/file-service.ts [hunk 8]
```diff
@@ -499,6 +580,13 @@ export class FileService {
         }));
     }
 
+    /**
+     * Tests if the given resource exists in the filesytem.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/browser/file-service.ts:584]:
```suggestion
     * Tests if the given resource exists in the filesystem.
```

File: packages/filesystem/src/browser/file-service.ts [hunk 9]
```diff
@@ -485,6 +559,13 @@ export class FileService {
         return fileStat;
     }
 
+    /**
+     * Try to resolve file information and metadata for all given resource.
+     * @param toResolve An array of all the resources (and corresponding resolvment options) that should be resolved.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/browser/file-service.ts:564]:
```suggestion
     * @param toResolve An array of all the resources (and corresponding resolvement options) that should be resolved.
```

**Review by vince-fugnitto — COMMENTED** [review]
There are a couple of typos, I haven't read through the actual documentation yet.

File: packages/filesystem/src/browser/file-service.ts (hunk 7 above)
Comment (tortmayr) [reply packages/filesystem/src/browser/file-service.ts:467]: Should be `resolvement`

Comment (tortmayr) [inline packages/filesystem/src/browser/file-service.ts:467]:
```suggestion
     * @param options  Options to customize the resolvement process.
```

**Review by vince-fugnitto — COMMENTED** [review]
@tortmayr please be sure to address the build issues (I believe formatting errors), and squash the commits into a single commit with the `sign-off`. There's no need to co-author me :) all I did was provide a review.

**Comment by sdirix** [comment]
Hi @vince-fugnitto thanks for the review! I addressed the remaining issue and squashed the commits.

File: packages/filesystem/src/browser/file-service.ts [hunk 10]
```diff
@@ -347,10 +401,23 @@ export class FileService {
...
     canHandleResource(resource: URI): boolean {
         return this.providers.has(resource.scheme);
     }
 
+    /**
+     * Tests if the service (i.e the {@link FileSystemProvider} registered for the given uri scheme) provides the given capability.
+     * @param resource `URI` of the resource to test.
+     * @param capability The required capability.
+     *
+     * @returns `true` if the resource can be handled and the required cabability can be provided.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/browser/file-service.ts:419]:
```suggestion
     * @returns `true` if the resource can be handled and the required capability can be provided.
```

File: packages/filesystem/src/common/files.ts [hunk 11]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+     */
     delete(resource: URI, opts: FileDeleteOptions): Promise<void>;
 
+    /**
+     * Rename a file or folder.
+     * @param from `URI` of the existing file or folder.
+     * @param to `URI` of the target location.
+     * @param opts Options to define if existing files should be overwritten.
+     */
+
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:626]: I believe there is an incorrectly added extra line.

File: packages/filesystem/src/common/files.ts [hunk 12]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
     rename(from: URI, to: URI, opts: FileOverwriteOptions): Promise<void>;
+
+    /**
+     * Optional function that has to be implemented by {@link FileSystemProviderWithFileFolderCopyCapability}.
+     * see {@link  FileSystemProviderWithFileFolderCopyCapability#copy}} for additional documentation.
```
Comment (vince-fugnitto) [inline packages/filesystem/src/common/files.ts:631]: There are extra _whitespaces_ after `@link` for the following methods.
Please ignore if these are done on purpose.

**Review by vince-fugnitto — COMMENTED** [review]
I reviewed the content of the documentation and it looks very good!
I identified a few minor comments (formatting, typos) to address.

File: packages/filesystem/src/browser/file-service.ts (hunk 10 above)
Comment (sdirix) [reply packages/filesystem/src/browser/file-service.ts:419]: Good catch!

File: packages/filesystem/src/common/files.ts (hunk 12 above)
Comment (sdirix) [reply packages/filesystem/src/common/files.ts:631]: I'm pretty sure that's a mistake ;)

**Comment by sdirix** [comment]
Addressed your comments and added some minor additional changes, adjusting whitespace and upper cases in beginning of sentences.

**Review by kittaakos — APPROVED** [review]
Thank you for taking care of Theia ❤️

File: packages/filesystem/src/common/files.ts [hunk 13]
```diff
@@ -558,85 +558,290 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+     * Create a new directory using the given resource uri.
+     * @param resource The `URI` of the new folder.
+     */
     mkdir(resource: URI): Promise<void>;
+
+    /**
+     * Retrieve the content of a given directory.
+     * @param resource The `URI` of the directory.
+     *
+     * @returns A map containing the {@link FileType} for each child resource (uri).
```
[the remark this answers is not in the document]
Comment (sdirix) [reply packages/filesystem/src/common/files.ts:609]: I'll check and come back to you :+1:

[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/filesystem/src/common/files.ts:609]: Thank you! VS Code uses `name`. I implemented an `FSProvider` based on your documentation yesterday, right before I merged it, and `readdir` did not work.

[the remark this answers is not in the document]
Comment (sdirix) [reply packages/filesystem/src/common/files.ts:609]: I believe you're correct. Debugging into a standard Theia application the existing providers only return file names here.

[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/filesystem/src/common/files.ts:609]: Thank you so much for checking it. Do you happen to have time to prepare a follow-up PR with the correction?

[the remark this answers is not in the document]
Comment (sdirix) [reply packages/filesystem/src/common/files.ts:609]: I'll open one in a short time :+1:

[the remark this answers is not in the document]
Comment (sdirix) [reply packages/filesystem/src/common/files.ts:609]: PR is here: https://github.com/eclipse-theia/theia/pull/8638.

Thanks for noticing this issue!

**Pull request merged**